### PR TITLE
Port commandline parsing to using GLib Application

### PR DESCRIPTION
**Pull request by Neui**
Done a bit more than I would've liked, but it is a step closer to being DBus-Activatable (#183).

More testing appreciated. An current bug I've found is that when opening an task via cli (`gtg://....`), closing the task (and thus the application) and open GTG normally, it would reopen the task. I am not sure whenever this happens with the old implementation, but it might be worth investigating later.

I tried to explain in the commits, but there you go if you're lazy:

---

Use GLibs cli parser and use application.open

The primarily goal was to use the GApplication command line parser, so
that an user can also access gtk and GLibs command line arguments.

However, the opening tasks part made this much bigger, since it is done
from the activate-signal, which is usually reserved to open without any
files. With the new system, the open-signal is used to open files, so it
had to be ported to over there.

Initializing the browser window wasn't possible in the startup-signal
(it crashes, and it seems you aren't supposed to do that), so this is
done by both activate- and open-signals.

An weird oddity I've found is that in "the" documentation for python,
the do_open just accepts 2 arguments: files and hint. However, I was
getting argument mismatch errors, and I've noticed that I got a number
as the second parameter, and by looking at the C docs, the second
parameter is the number of files.
Since files is a python list, this isn't needed, but checked anyway,
just in case.
https://lazka.github.io/pgi-docs/Gio-2.0/classes/Application.html#Gio.Application.do_open
https://developer.gnome.org/gio/2.66/GApplication.html#GApplication-open

The files in the open-signal are GFiles, which seem to change the URI a
tiny bit by adding a final slash (gtg://a → gtg://a/). Because of this,
the old code wouldn't find the task, so I went ahead and parsed the URI
completely using urllib.parse.
GLib also has URI parsing, but it seems they aren't exposed to python.

Also, GLib also translates the help strings if they are translated.
However, I needed to put GTG/gtg.in manually in po/POTFILES.in, but even
then, xgettext defaulted to using C instead of python (rightfully).
An workaround would be creating an symlink gtg.py.in and inputting that
to xgettext, but I didn't wanted to deal more with this, so I didn't
bother to continue.

---

scripts/debug.sh: Support passing directly to gtg

Useful to input custom arguments or tasks directly to gtg.
Note that from what I understand, it would get the options up to the
last non-option and pass it to gtg, but I think it isn't too bad for a
debug script.

---

prefix-gtg.sh.in: Properly "escape" arguments

Previously, the launch.sh (scripts/debug.sh) script would pass a title
like -t "Dev GTG", but gtg would read it as ["-t", "Dev", "GTG"].
This commit fixes this.

From `man bash`:
>      @      Expands  to the positional parameters, starting from one.
>             When the expansion  occurs  within  double  quotes,  each
>             parameter  expands  to a separate word.  That is, "$@" is
>             equivalent to "$1" "$2" ...  If the double-quoted  expan‐
>             sion  occurs  within  a  word, the expansion of the first
>             parameter is joined with the beginning part of the origi‐
>             nal  word,  and  the  expansion  of the last parameter is
>             joined with the last part of  the  original  word.   When
>             there are no positional parameters, "$@" and $@ expand to
>             nothing (i.e., they are removed).

---

Try to quit when just opening files

A side effect of moving to Gtk.Application with the current MainWindow
dependency is that detecting whenever to quit the application is
harder.

Assigning an application to an window keeps the application open when
the window exists. So by not assigning it, and when just tasks were
open, upon closing them quits the application as expected.
However, when activated, it sets the application and continues as
normal.

I don't really like this solution, but is pretty easy to read (and
hopefully to understand).
I hope that this doesn't cause bugs around quitting application when all
windows are closed.

### CODE REVIEW

**Comment by Neui** [comment]
Removed one commit and changed the first one because I noticed when you pass tasks to open it'll also open the main window in current master, and thus changed to reflect that.

File: GTG/gtg.in [hunk 1]
```diff
@@ -63,40 +65,31 @@ from GTG.core import info
...
-    """Parse arguments from the command line."""
-
-    parser = argparse.ArgumentParser()
-
-    parser.add_argument('-v', '--version', help='Show program version',
-                        action="store_true")
-
-    parser.add_argument('-d', '--debug', help='Enable debug output',
-                        action='store_true')
+def handle_local_options(application, options):
```
Comment (diegogangl) [inline GTG/gtg.in:68]: A docstring here would be nice

File: GTG/gtk/application.py [hunk 2]
```diff
@@ -76,16 +74,11 @@ class Application(Gtk.Application):
     delete_task_dialog = None
     edit_tag_dialog = None
 
-    def __init__(self, app_id, debug):
+    def __init__(self, app_id=None):
```
Comment (diegogangl) [inline GTG/gtk/application.py:77]: Why make app_id optional?

File: GTG/gtk/application.py [hunk 3]
```diff
@@ -121,19 +114,49 @@ def __init__(self, app_id, debug):
     def do_activate(self):
         """Callback when launched from the desktop."""
 
+        self.init_shared()
+
+        log.debug("Application activation finished")
+
+    def do_open(self, files, n_files, hint):
+        # Note: n_files isn't in the python doc, but is in the c doc and is required
```
Comment (diegogangl) [inline GTG/gtk/application.py:122]: I'd rather leave this information in the commit. Nobody is calling this function directly and the docs might change

File: GTG/gtk/application.py [hunk 4]
```diff
@@ -121,19 +114,49 @@ def __init__(self, app_id, debug):
...
+                    parsed = urllib.parse.urlparse(uri)
+                    task_id = parsed.netloc
+                    log.debug(f'Opening task {task_id}')
+                    self.open_task(task_id)
+            else:
+                log.info(f"Unknown task to open: {file.get_uri()}")
+
+        log.debug("Application opening finished")
+
+    def init_shared(self):
```
Comment (diegogangl) [inline GTG/gtk/application.py:146]: Missing a docstring here

File: GTG/gtk/application.py [hunk 5]
```diff
@@ -533,3 +540,15 @@ def do_shutdown(self):
         self.req.save_datastore(quit=True)
 
         Gtk.Application.do_shutdown(self)
+
+    # --------------------------------------------------------------------------
+    # MISC
+    # --------------------------------------------------------------------------
+
+    def set_debug_flag(self, debug):
```
Comment (diegogangl) [inline GTG/gtk/application.py:548]: Missing docstring here

File: GTG/gtg.in [hunk 6]
```diff
@@ -63,40 +65,31 @@ from GTG.core import info
...
-        if args.version:
-            print("GTG (Getting Things GNOME!)", info.VERSION)
-            print()
-            print("For more information:", info.URL)
-            sys.exit(0)
-
-        if args.title is not None:
-            info.NAME = args.title
+        def N_(message): # Mark as to be translated for gettext
+            return message
```
Comment (diegogangl) [inline GTG/gtg.in:92]: Convert the comment into a docstring

**Review by diegogangl — CHANGES_REQUESTED** [review]
LGTM, everything works as expected. Some minor nitpicks only

File: GTG/gtk/application.py (hunk 2 above)
Comment (Neui) [reply GTG/gtk/application.py:77]: I don't remember anymore... I'll make it mandatory

File: GTG/gtg.in (hunk 1 above)
Comment (Neui) [reply GTG/gtg.in:68]: Done

File: GTG/gtg.in (hunk 6 above)
Comment (Neui) [reply GTG/gtg.in:98]: Done

File: GTG/gtk/application.py (hunk 4 above)
Comment (Neui) [reply GTG/gtk/application.py:145]: Done

File: GTG/gtk/application.py (hunk 5 above)
Comment (Neui) [reply GTG/gtk/application.py:552]: Done

**Comment by Neui** [comment]
I've force-pushed and did the stuff and changed the first commit a bit for more info. I'll let you resolve the "conversations".
Actually, I've noticed I could place an extra call on `do_activate` to call `self.browser.present()`, because if the application is already open but minimized, and the user tries to open the application again, moving the window to the foreground seems better than just showing on the first time. I'll put the additional line into this PR if you're OK.

**Comment by diegogangl** [comment]
> Actually, I've noticed I could place an extra call on do_activate to call self.browser.present(), because if the application is already open but minimized, and the user tries to open the application again, moving the window to the foreground seems better than just showing on the first time. I'll put the additional line into this PR if you're OK.

Sure, go ahead

**Comment by diegogangl** [comment]
Merged, thanks!

**Pull request merged**